packages/k8s.contrib.crd/internal/ModuleGenerator.pkl

```diff
@@ -0,0 +1,366 @@
+              else new TypeNode.NullableTypeNode { typeNode = underlyingType }
+        }
+        defaultValue =
+          if (typeAnnotation.type is TypeNode.UnionTypeNode
+          && (typeAnnotation.type as TypeNode.UnionTypeNode).members[0] is TypeNode.StringLiteralTypeNode
+          )
+            // If the first member is a string literal type, explicitly set the default to Undefined().
+            // We need to do this because the implicit default of a union type is the default of the first member.
+            // However, this is an incorrect way to represent enums in JSON Schema, because there should be no
+            // bias towards any of the values.
+            new ExpressionNode.MemberAccessExpressionNode {
+              identifier {
+                value = "Undefined"
+              }
+              // empty arguments makes this a function call
+              arguments {}
+            }
+          else null
+      }
+    }
+  }
+
+/// Generates a [ClassNode] from a [CRDJsonSchema].
+local function generatePklClass(schema: JsonSchema, className: Type, typeNames: Type.TypeNames): ClassNode =
+  new {
+    docComment = getDocComment(schema, "class")
+    classHeader {
+      name {
+        value = className.name
+      }
+    }
+    properties = generateClassBody(schema, typeNames)
+  }
+
+function isClassLike(schema: JsonSchema.Schema): Boolean =
+  if (schema is Boolean)
+    false
+  else
+    schema.properties != null
+
+/// Determine the name of a type.
+///
+/// Try to use the parent property's name as part of the class name in case of conflict.
+/// If already at the root, add a number at the end.
+local function determineTypeName(path: List<String>, candidateName: String, existingTypeNames: Set<Type>, index: Int): Type =
+  if (existingTypeNames.contains(utils.pascalCase(candidateName)))
+    if (path.isEmpty)
+      determineTypeName(path, candidateName + index.toString(), existingTypeNames, index + 1)
+    else
+      determineTypeName(path.dropLast(1), utils.pascalCase(path.last.capitalize()) + utils.pascalCase(candidateName), existingTypeNames, index)
+  else
+    new { name = utils.pascalCase(candidateName); moduleName = module.moduleName }
+
+/// The schemas that should be rendered as classes.
+///
+/// Classes get rendered for any subschema that has [JsonSchema.properties] defined, and does not show up in converters
+local classSchemas: Type.TypeNames =
+  utils._findMatchingSubSchemas(filteredRootSchema, List(), (elem) -> elem != filteredRootSchema && isClassLike(elem))
+    .filter((path, _) -> !pathPrefixes(path).any((prefix) -> converters.containsKey(prefix))) // path or prefix are not explicitly in converters
+    .entries
+    .fold(Map(), (accumulator: Type.TypeNames, pair) ->
+    let (path = pair.first)
+      let (schema = pair.second)
+        let (typeName = determineTypeName(path, path.lastOrNull?.capitalize() ?? "Item", accumulator.values.toSet(), 0))
+          accumulator.put(schema, typeName)
```

Comment:
There's quite a bit of packing/unpacking going on here; are you _intentionally_ being quiet in the case of repeated map keys?